language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/ContextTypeResolver.java

```diff
@@ -0,0 +1,262 @@
+import io.ballerina.compiler.api.symbols.RecordTypeSymbol;
+import io.ballerina.compiler.api.symbols.Symbol;
+import io.ballerina.compiler.api.symbols.SymbolKind;
+import io.ballerina.compiler.api.symbols.TypeDefinitionSymbol;
+import io.ballerina.compiler.api.symbols.TypeDescKind;
+import io.ballerina.compiler.api.symbols.TypeSymbol;
+import io.ballerina.compiler.syntax.tree.AnnotationNode;
+import io.ballerina.compiler.syntax.tree.AssignmentStatementNode;
+import io.ballerina.compiler.syntax.tree.CaptureBindingPatternNode;
+import io.ballerina.compiler.syntax.tree.FunctionDefinitionNode;
+import io.ballerina.compiler.syntax.tree.IdentifierToken;
+import io.ballerina.compiler.syntax.tree.IndexedExpressionNode;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.NodeTransformer;
+import io.ballerina.compiler.syntax.tree.QualifiedNameReferenceNode;
+import io.ballerina.compiler.syntax.tree.ReturnTypeDescriptorNode;
+import io.ballerina.compiler.syntax.tree.SimpleNameReferenceNode;
+import io.ballerina.compiler.syntax.tree.SpecificFieldNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.TypeDescriptorNode;
+import io.ballerina.compiler.syntax.tree.VariableDeclarationNode;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.common.utils.SymbolUtil;
+import org.ballerinalang.langserver.common.utils.completion.QNameReferenceUtil;
+import org.ballerinalang.langserver.commons.BallerinaCompletionContext;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+import javax.annotation.Nonnull;
+
+/**
+ * This visitor is used to resolve a type of a given context.
+ * For example, consider the following example,
+ * <pre>
+ *     function sayHello() returns int {
+ *         ...
+ *         Person p = {[cursor]}
+ *         ...
+ *     }
+ * </pre>
+ * For the given cursor position the resolved type is Person. At the usage of this resolver, we provide the mapping
+ * expression and the resolver traverse the parents to capture the possible type.
+ * <p>
+ * Consider the following
+ *
+ * <pre>
+ *     function getPerson() PersonRecord {
+ *         ...
+ *         return {[cursor]}
+ *     }
+ * </pre>
+ * For the above example the visible type is resolved by the return type of the function definition, which is
+ * PersonRecord
+ *
+ * @since 2.0.0
+ */
+public class ContextTypeResolver extends NodeTransformer<Optional<TypeSymbol>> {
+    private final BallerinaCompletionContext context;
+    private final List<Node> visitedNodes = new ArrayList<>();
+
+    public ContextTypeResolver(BallerinaCompletionContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(SimpleNameReferenceNode node) {
+        Optional<Symbol> variableSymbol = this.getSymbolByName(node.name().text());
```

Comment:
Can't we use `SemanticModel.symbol(node)` instead of this. Then we can cover  `transform(QualifiedNameReferenceNode node)` without ambiguity as well ?

Reply:
The qualified name reference is implemented to capture the type definitions specifically, we bypass the ambiguity of the annotation symbol in the particular node visitor. For the fall back if in any case there are two members we return empty until we handle it properly. Also, the variable name based capturing is used to skip capturing the type reference type symbol instead of the type definition.